subsys/sip_svc/sip_svc.c

```diff
@@ -0,0 +1,898 @@
+ *                                             V
+ * ------------------------------------------------------
+ * EL2/EL3                                      ^
+ * firmware                                     |
+ *                                   Return same Transaction ID
+ * ------------------------------------------------------
+ *
+ */
+
+#define LOG_LEVEL CONFIG_ARM_SIP_SVC_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(sip_svc);
+
+#include <zephyr.h>
+#include <string.h>
+#include <device.h>
+#include <zephyr/sip_svc/sip_svc.h>
+#include "sip_svc_id_mgr.h"
+
+__weak bool sip_svc_plat_func_id_valid(uint32_t command, uint32_t func_id)
+{
+	if (command <= SIP_SVC_PROTO_CMD_MAX)
+		return true;
+	else
+		return false;
+}
+
+__weak uint32_t sip_svc_plat_format_trans_id(uint32_t client_idx,
+				uint32_t trans_idx)
+{
+	return (((client_idx & 0xF) << 4) | (trans_idx & 0xF));
+}
+
+__weak uint32_t sip_svc_plat_get_trans_idx(uint32_t trans_id)
+{
+	return (trans_id & 0xF);
+}
+
+__weak void sip_svc_plat_update_trans_id(struct sip_svc_request *request,
+				uint32_t trans_id)
+{
+}
+
+__weak void sip_svc_plat_free_async_memory(struct sip_svc_request *request)
+{
+}
+
+__weak int sip_svc_plat_async_res_req(unsigned long *a0, unsigned long *a1,
+				unsigned long *a2, unsigned long *a3,
+				unsigned long *a4, unsigned long *a5,
+				unsigned long *a6, unsigned long *a7,
+				char *buf, size_t size)
+{
+	return -ENOTSUP;
+}
+
+__weak int sip_svc_plat_async_res_res(struct arm_smccc_res *res,
+				char *buf, size_t *size,
+				uint32_t *trans_id)
+{
+	return -ENOTSUP;
+}
+
+
+__weak uint32_t sip_svc_plat_get_error_code(struct arm_smccc_res *res)
+{
+	return -ENOTSUP;
+}
+
+#if CONFIG_ARM_SIP_SVC_HAS_HVC_CALLS
+static void __invoke_fn_hvc(unsigned long function_id,
+					  unsigned long arg0,
+					  unsigned long arg1,
+					  unsigned long arg2,
+					  unsigned long arg3,
+					  unsigned long arg4,
+					  unsigned long arg5,
+					  unsigned long arg6,
+					  struct arm_smccc_res *res)
+{
+	arm_smccc_hvc(function_id, arg0, arg1, arg2, arg3,
+		      arg4, arg5, arg6, res);
+}
+#endif
+
+#if CONFIG_ARM_SIP_SVC_HAS_SMC_CALLS
+static void __invoke_fn_smc(unsigned long function_id,
+					  unsigned long arg0,
+					  unsigned long arg1,
+					  unsigned long arg2,
+					  unsigned long arg3,
+					  unsigned long arg4,
+					  unsigned long arg5,
+					  unsigned long arg6,
+					  struct arm_smccc_res *res)
+{
+	arm_smccc_smc(function_id, arg0, arg1, arg2, arg3,
+		      arg4, arg5, arg6, res);
+}
+#endif
+
+static struct sip_svc_controller controller[] = {
+#if CONFIG_ARM_SIP_SVC_HAS_SMC_CALLS
+	{ .init = false, .method = "smc", .invoke_fn = __invoke_fn_smc },
+#endif
+#if CONFIG_ARM_SIP_SVC_HAS_HVC_CALLS
+	{ .init = false, .method = "hvc", .invoke_fn = __invoke_fn_hvc },
+#endif
+	{.init = false, .method = "", .invoke_fn = NULL },
+};
+
+static uint32_t sip_svc_generate_c_token(void)
+{
+	uint32_t c_token = k_cycle_get_32();
```

Comment:
Not for sure if it is your case, just a kind reminder. Some platform has a low freq timer, when you continuously call `k_cycle_get_32()`, you might get 2 or more same tokens. If it doesn't matter, just ignore this:)